source/particle/interpolator/nearest_neighbor.cc

```diff
@@ -0,0 +1,163 @@
+#include <aspect/postprocess/particles.h>
+#include <aspect/simulator.h>
+
+#include <deal.II/grid/grid_tools.h>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Interpolator
+    {
+      template <int dim>
+      std::vector<std::vector<double> >
+      NearestNeighbor<dim>::properties_at_points(const std::multimap<types::LevelInd, Particle<dim> > &particles,
+                                                 const std::vector<Point<dim> > &positions,
+                                                 const ComponentMask &selected_properties,
+                                                 const typename parallel::distributed::Triangulation<dim>::active_cell_iterator &cell) const
+      {
+        const Postprocess::Particles<dim> *particle_postprocessor = this->template find_postprocessor<Postprocess::Particles<dim> >();
+
+        typename parallel::distributed::Triangulation<dim>::active_cell_iterator found_cell;
+
+        if (cell == typename parallel::distributed::Triangulation<dim>::active_cell_iterator())
+          {
+            // We can not simply use one of the points as input for find_active_cell_around_point
+            // because for vertices of mesh cells we might end up getting ghost_cells as return value
+            // instead of the local active cell. So make sure we are well in the inside of a cell.
+            Assert(positions.size() > 0,
+                   ExcMessage("The particle property interpolator was not given any "
+                              "positions to evaluate the particle properties at."));
+
+            const Point<dim> approximated_cell_midpoint = std::accumulate (positions.begin(), positions.end(), Point<dim>())
+                                                          / static_cast<double> (positions.size());
+
+            found_cell =
+              (GridTools::find_active_cell_around_point<> (this->get_mapping(),
+                                                           this->get_triangulation(),
+                                                           approximated_cell_midpoint)).first;
+          }
+        else
+          found_cell = cell;
+
+        const types::LevelInd cell_index = std::make_pair(found_cell->level(),found_cell->index());
+
+        const std::pair<typename std::multimap<types::LevelInd, Particle<dim> >::const_iterator,
+              typename std::multimap<types::LevelInd, Particle<dim> >::const_iterator> particle_range =
+                (cell->is_locally_owned())
+                ?
+                particles.equal_range(cell_index)
+                :
+                particle_postprocessor->get_particle_world().get_ghost_particles().equal_range(cell_index);
+
+        const unsigned int n_particles = std::distance(particle_range.first,particle_range.second);
+        const unsigned int n_particle_properties = particles.begin()->second.get_properties().size();
+
+        std::vector<std::vector<double> > point_properties;
```

Comment:
You already know how large this vector of vectors is going to be, and you can do all of the memory allocation here, instead of inside of the inner loops below the following way:
```
std::vector<double> temp(n_particle_properties,0.0);
point_properties (n_positions.size(),temp);
```
Then below you can always fill `point_properties[pos_idx][i]` instead of calling `push_back()`. This should speed things up significantly (I do not know how fast or slow this algorithm currently is, likely faster than cell average, but no reason to waste efficiency).